File: Search/src/main/java/bfs/Solution433.java
package bfs;

import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Queue;

public class Solution433 {
    public int minMutation(String start, String end, String[] bank) {
        char[] genes = { 'A', 'C', 'G', 'T' };
        HashMap<String, Integer> depth = new HashMap<>();

        HashSet<String> bankSet = new HashSet<>(Arrays.asList(bank));
        Queue<String> q = new ArrayDeque<>();

        if (!bankSet.contains(end)) {
            return -1;
        }

        depth.put(start, 0);
        q.add(start);

        while (!q.isEmpty()) {
            String first = q.poll();

            for (int i = 0; i < 8; i++) {
                for (char gene : genes) {
                    if (first.charAt(i) == gene) {
                        continue;
                    }
                    String newFirst = first.substring(0, i) + gene + first.substring(i + 1, first.length());

                    if (!bankSet.contains(newFirst)) {
                        continue;
                    }
                    if (depth.containsKey(newFirst)) {
                        continue;
                    }

                    depth.put(newFirst, depth.getOrDefault(first, 0) + 1);

                    q.add(newFirst);

                    if (newFirst.equals(end)) {
                        return depth.get(newFirst);
                    }
                }
            }

        }

        return -1;

    }

    public static void main(String[] args) {
        Solution433 s = new Solution433();

        String start = "AACCGGTT";
        String end = "AAACGGTA";
        String[] bank = { "AACCGGTA", "AACCGCTA", "AAACGGTA" };
        int res = s.minMutation(start, end, bank);
        System.out.print(res);

    }
}

File: Search/src/main/python/bfs/solution433.py
from typing import List


class Solution:
    def minMutation(self, start: str, end: str, bank: List[str]) -> int:
        genes = ["A", "C", "G", "T"]
        depth = {start: 0}

        if end not in bank:
            return -1

        queue = []

        queue.append(start)

        while len(queue) != 0:
            s = queue[0]
            queue.pop(0)  # pop 0 是从前面去除

            for i in range(8):
                for gene in genes:
                    if s[i] == gene:
                        continue
                    ns = s[:i] + gene + s[i + 1 :]
                    if ns not in bank:
                        continue
                    # if depth.get(ns):  # 不能用这个来判断是否有这个key， 因为有0
                    if ns in depth:
                        continue

                    depth[ns] = depth[s] + 1
                    queue.append(ns)

                    if ns == end:
                        return depth[ns]

        return -1


# start = "AACCGGTT"
# end = "AAACGGTA"
# bank = ["AACCGGTA", "AACCGCTA", "AAACGGTA"]

# start = "AAAAAAAA"
# end = "CCCCCCCC"
# bank = [
#     "AAAAAAAA",
#     "AAAAAAAC",
#     "AAAAAACC",
#     "AAAAACCC",
#     "AAAACCCC",
#     "AACACCCC",
#     "ACCACCCC",
#     "ACCCCCCC",
#     "CCCCCCCA",
#     "CCCCCCCC",
# ]
start = "AAAACCCC"
end = "CCCCCCCC"
bank = [
    "AAAACCCA",
    "AAACCCCA",
    "AACCCCCA",
    "AACCCCCC",
    "ACCCCCCC",
    "CCCCCCCC",
    "AAACCCCC",
    "AACCCCCC",
]

s = Solution()
res = s.minMutation(start, end, bank)

print(res)